docs/src/modules/java/pages/ai-coding-assistant-guidelines.adoc

```diff
@@ -2,45 +2,114 @@
+* A domain record should be immutable.
+** Make a copy of collections when updating them.
+** To make single-field mutation easier, prefer the functional builder-style syntax, e.g. `withLocation(x)` that creates a new instance with the new field value rather than explicitly setting the location field
+
 
 == Entities
 
-- Commands can be defined as records inside the entity and they don't have to have a common interface.
-- Command handlers should be implemented in the entity and not in the domain object. The domain object should have methods for business logic validations but it should not handle commands directly or be concerned with entity effects.
-- Command handlers that make updates without returning any information should return `akka.Done` for successful responses and `effects().error()` for validation errors. Use static import for `akka.Done.done()` method.
-- Events should be defined in the domain package and the events for an entity should have a common sealed interface, and define `@TypeName` for serialization.
-- `applyEvent` method should never return null, return the current state or throw an exception.
-- Key Value Entities, Event Sourced Entities, Workflows can accept only single method parameter, wrap multiple parameters in a record class.
+* Command handlers only ever accept a single parameter and return effects. If a command handler needs to accept more than one parameter, those parameters should be wrapped inside a single command message.
+* The command handler parameter and reply should be defined as inner record of an entity, but sometimes it makes sense to place them in the `domain` package together with other data transfer and state records. Parameter and reply can also be a plain String, Java primitive or class.
+* Commands should be defined using an imperative naming convention, described as instructions of some task to be performed. For example, a command requesting a shopping cart be checked out would be `ShoppingCartEntity.Checkout`.
+* Command handlers are to be implemented in the entity and not in the state or domain object. Command handlers should invoke the domain object’s validation methods and methods that encapsulate the core business logic. The entity’s command handler is responsible for choosing the appropriate effect(s) to return. The domain object should not return effects or a list of events. This keeps the domain model clean of infrastructural concerns and makes it easy to unit test in isolation.
+* Command handlers that make updates without returning any information should return `akka.Done` for successful responses and `effects().error()` for validation errors.
+* State of an entity should be in the `domain` package, also known as domain object.
+* The `applyEvent` override should be a pure function transferring data from the event to the state and nothing else. It should will never fail, because validations should have been made before persisting the event.
```

Comment:
```suggestion
* The `applyEvent` override should be a pure function transferring data from the event to the state and nothing else. It should never fail, because validations should have been made before persisting the event.
```